Library/Homebrew/downloadable.rb

```diff
@@ -0,0 +1,132 @@
+# typed: true
+# frozen_string_literal: true
+
+require "url"
+require "checksum"
+
+# @api private
+class Downloadable
+  extend T::Sig
+
+  include Context
+
+  sig { returns(T.nilable(URL)) }
+  attr_reader :url
```

Comment:
When would a URL be `nil`?

Reply:
Ideally it wouldn't but it can be due to the way that `url` may not be set immediately.

For resources: URL is defined later in a DSL in a `do block` but the nature of that means it may not actually be called and could be nil - we had various `url.nil?` checks already in `resource.rb` prior to this PR.
For cask downloads: we have a concept of lazily loaded URLs where the URL can only be calculated at install time. Plus casks allow basically anything to be nil at the moment.

I spent a fair bit of time fighting that but I can try again, potentially by ditching the instance variable altogether and deferring a getter to subclasses.

Reply:
No, all good, leave it as-is, all makes sense and just curious 👍🏻